test/unit/math/prim/prob/discrete_range_test.cpp

```diff
@@ -0,0 +1,73 @@
+#include <stan/math/prim.hpp>
+#include <boost/math/distributions.hpp>
+#include <boost/random/mersenne_twister.hpp>
+#include <gtest/gtest.h>
+#include <limits>
+#include <vector>
+
+TEST(ProbDistributionsDiscreteRange, error_check) {
+  using stan::math::discrete_range_rng;
+  boost::random::mt19937 rng;
+
+  double nan = std::numeric_limits<double>::quiet_NaN();
+  double inf = std::numeric_limits<double>::infinity();
+
+  Eigen::VectorXd lower(3);
+  Eigen::VectorXd upper(3);
+
+  lower << 5.0, 11.0, -15.0;
+  upper << 5.0, 15.0, -10.0;
+  EXPECT_NO_THROW(discrete_range_rng(lower, upper, rng));
```

Comment:
This should be illegal, so it looks like that Boost thing lets integers get assigned.  So the signature should prevent this with an enable_if on the template parameters.  The tests need to be updated then to use integers.

Reply:
No other distribution uses any `enable_if` or `require_` on the template parameters or elsewhere. So my understanding was that this is  enforced by the signatures we expose in the compiler. I can try to concoct something, but it'd be a first in the codebase. I'll update the tests.

Reply:
The way to enforce that all arguments are integers or integer arrays, add a final template parameter:

```
typename = std::enable_if_t<std::is_integral<scalar_type_t<T1>>::value
                            && std::is_integral<scalar_type_t<T2>>::value
                            && std::is_integral<scalar_type_t<T3>>::value>
```

where `scalar_type_t` is a Stan meta function.

This isn't strictly necessary, but it should be clear in the doc that the use is intended to only be for integers.

Reply:
Just to be clear here, our `binomial_lpmf` and our `hypergeometric_lpmf` are broken in that they'll accept real-valued arguments.  I'm going to open a separate issue to try to deal with some of this as it'll be easy to call them erroneously as they're written with real values, but those don't make sense (they don't normalize properly).